File: Week-2-April-8th-14th/Last-Stone-Weight/last-stone-weight.py
# Time complexityL O(n^2) ... Space complexity: O(1)
class Solution:
    def getMax(self, stones):
        self.firstMax = self.secMax = -float("inf")
        self.firstInd = -1
        
        for i in range(len(stones)):
            if stones[i] > self.firstMax:
                self.firstMax = stones[i]
                self.firstInd = i
        for i in range(len(stones)):
            if stones[i] > self.secMax and i != self.firstInd:
                self.secMax = stones[i]
        return [self.firstMax,self.secMax]
    
    def lastStoneWeight(self, stones: List[int]) -> int:
        while len(stones) > 1: # O(n/2)
            maxVals = self.getMax(stones)
            tmp = abs(maxVals[0] - maxVals[1])
            stones.remove(maxVals[0]) #O(n)
            stones.remove(maxVals[1])
            if tmp != 0:
                stones.append(tmp)
        if len(stones) == 1:
            return stones[0]
        return 0

File: Week-2-April-8th-14th/Last-Stone-Weight/last-stone-weight.cpp
// Method 1
// Time complexity: O(n^2logn)
// Space complexity: O(1)
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {        
        while(stones.size() > 1){ // O(n/2)
            sort(stones.begin(), stones.end()); // O(nlogn)
            int len = stones.size();
            int tmp = abs(stones[len-1] - stones[len-2]);
            stones.pop_back(); 
            stones.pop_back();
            if(tmp != 0){
                stones.push_back(tmp);
            }
        }
        if(stones.size() == 1){
            return stones[0];
        }
        return 0;
    }
};

// Method 2: Using priority_queue
// Time complexity: O(nlogn)
// Space complexity: O(n)
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq(stones.begin(), stones.end()); // O(n)
        
        while(pq.size() > 1){ // O(n)
            int firstHeaviest = pq.top(); // O(1)
            pq.pop(); // O(logn)
            int secondHeaviest = pq.top();
            pq.pop();
            
            int diff = firstHeaviest - secondHeaviest;
            if(diff != 0) {
                pq.push(diff);
            }
        }
        return pq.empty() ? 0 : pq.top();
    }
};
